stdlib/public/core/BridgeObjectiveC.swift

```diff
@@ -640,3 +641,200 @@ public func _getObjCTypeEncoding<T>(_ type: T.Type) -> UnsafePointer<Int8> {
+
+@usableFromInline
+protocol _Unwrappable {
+  func unwrap() -> Any?
+}
+
+extension Optional: _Unwrappable {
+  func unwrap() -> Any? {
+    return self
+  }
+}
+
+// This is a best-effort tripmine for detecting the situation
+// (which should never happen) of Swift._SwiftValue and
+// Foundation._SwiftValue/Foundation.NSNull being used 
+// in the same process.
+
+@usableFromInline
+internal enum _SwiftValueFlavor: Equatable {
+  case stdlib
+  case foundation
+}
+
+@usableFromInline
+func _currentSwiftValueFlavor() -> _SwiftValueFlavor {
+  if _typeByName("Foundation._SwiftValue") as? _NSSwiftValue.Type != nil {
+    return .foundation
+  } else {
+    return .stdlib
+  }
+}
+
+@usableFromInline
+internal var _selectedSwiftValueFlavor: _SwiftValueFlavor = _currentSwiftValueFlavor()
+
+@usableFromInline
+internal func _assertSwiftValueFlavorIsConsistent() {
+  assert(_selectedSwiftValueFlavor == _currentSwiftValueFlavor())
+}
+
+@usableFromInline
+internal var _nullPlaceholder: AnyObject {
+  _assertSwiftValueFlavorIsConsistent()
+  if let foundationType = _typeByName("Foundation._SwiftValue") as? _NSSwiftValue.Type {
+    return foundationType.null
+  } else {
+    return _SwiftValue.null
+  }
+}
+
+@usableFromInline
+func _makeSwiftValue(_ value: Any) -> AnyObject {
+  _assertSwiftValueFlavorIsConsistent()
+  if let foundationType = _typeByName("Foundation._SwiftValue") as? _NSSwiftValue.Type {
+    return foundationType.init(value)
+  } else {
+    return _SwiftValue(value)
+  }
+}
+
+/// Bridge an arbitrary value to an Objective-C object.
+///
+/// - If `T` is a class type, it is always bridged verbatim, the function
+///   returns `x`;
+///
+/// - otherwise, if `T` conforms to `_ObjectiveCBridgeable`,
+///   returns the result of `x._bridgeToObjectiveC()`;
+///
+/// - otherwise, we use **boxing** to bring the value into Objective-C.
+///   The value is wrapped in an instance of a private Objective-C class
+///   that is `id`-compatible and dynamically castable back to the type of
+///   the boxed value, but is otherwise opaque.
+///
+/// COMPILER_INTRINSIC
+@inlinable // FIXME(sil-serialize-all)
+public func _bridgeAnythingToObjectiveC<T>(_ x: T) -> AnyObject {
+  var done = false
+  var result: AnyObject!
+  
+  var source: Any = x
+  
+  if let dynamicSource = _extractDynamicValue(x) {
+    result = dynamicSource as AnyObject
+    done = true 
+  }
+  
+  if !done, let wrapper = source as? _Unwrappable {
+    if let value = wrapper.unwrap() {
+      result = value as AnyObject
+    } else {
+      result = _nullPlaceholder
+    }
+    
+    done = true
+  }
+
+  if !done {
+    if type(of: source) as? AnyClass != nil {
+      result = unsafeBitCast(x, to: AnyObject.self)
+    } else if let object = _bridgeToObjectiveCUsingProtocolIfPossible(source) {
+      result = object
+    } else {
+      result = _makeSwiftValue(source)
+    }
+  }
+  
+  return result
+}
```

Comment:
I feel like this would be best written in C++ to have full access to the low-level metadata APIs. Keeping the entire implementation would be cleaner, and probably lead to more understandable code, than doing the oblique things you need to approximate this in Swift slowly and opaquely, such as sprinkling new protocols like `_Unwrappable` for peeling optionals or new small bits of runtime/stdlib SPI like `_extractDynamicValue` for peeling existentials. With `Optional` we also need to be careful that boxing preserves the difference between different levels of `.none` nesting in nested `Optional`s—for `T??`, `.none` and `.some(.none)` should be able to round-trip through a box as distinct values.

Reply:
Part of why I went with Swift over C++ was the issue of correctness of memory management; is there documentation on the expectations and SPI? My understanding is that with the current work the documentation in the repository may no longer apply.

Reply:
The main thing that would've changed since then is that Swift functions now take arguments +0 by default (and we have macros to try to handle both +1 and +0 conventions as a transitional/fallback mechanism, though I'm not sure whether that's important for new code, cc @gottesmm). The Darwin implementation of bridging lives largely in the C++ runtime, and we have fairly extensive runtime tests that you should be able to gradually enable that check for object lifetime issues.